storage/diskmetricstore.go

```diff
@@ -425,7 +430,25 @@ func (dms *DiskMetricStore) restore() error {
 	}
 	defer f.Close()
 	d := gob.NewDecoder(f)
-	return d.Decode(&dms.metricGroups)
+	if err := d.Decode(&dms.metricGroups); err == nil {
+		return nil
+	}
+	// Convert legacy disk format. TODO(beorn7): Remove prior to v1 release.
+	level.Info(dms.logger).Log("msg", "could not decode persistence file, trying legacy v0.9 format")
+	legacyMetricGroups := legacyGroupingKeyToMetricGroup{}
+	// Need to rewind file and create new decoder.
+	if _, err := f.Seek(0, 0); err != nil {
+		return err
+	}
+	d = gob.NewDecoder(f)
+	if err := d.Decode(&legacyMetricGroups); err != nil {
+		return err // That's a real failure then, unrelated to the format change.
+	}
+	for _, mg := range legacyMetricGroups {
+		dms.metricGroups[groupingKeyFor(mg.Labels)] = mg
+	}
+	level.Info(dms.logger).Log("msg", "conversion from v0.9 legacy format successful, next persisting will happen in current format")
+	return nil
```

Comment:
So you don't encapsulate this code into a restore legacy function because will anyway be removed soon? (together with v1 release)

Reply:
Kind of. I guess it could be extracted as a separate function. But with passing in an open file handle and such, it would be a quite complex call. I think the inline code is just easier to understand. The fact that we'll remove the code very soon is only important as it removes the code maintenance argument.